packages/executor/src/index.ts

```diff
@@ -1 +1,118 @@
+  const { positionId, owner, marketAddress } = position
+
+  try {
+    const privateKeys = process.env.PRIVATE_KEYS?.split(',') ?? []
+
+    const provider = new ethers.providers.JsonRpcProvider(process.env.RPC_URLS?.split(',')[0])
+    const marketContract = new ethers.Contract(marketAddress, market_abi, provider)
+
+    for (const privateKey of privateKeys) {
+      const wallet = new ethers.Wallet(privateKey, provider)
+
+      const [nonce, pendingNonce] = await Promise.all([
+        provider.getTransactionCount(wallet.address, 'latest'),
+        provider.getTransactionCount(wallet.address, 'pending'),
+      ])
+
+      if (nonce !== pendingNonce) {
+        log(chalk.red(`Nonce conflict for wallet ${wallet.address}, trying next...`))
+        continue
+      }
+
+      log(
+        `${chalk.green('Starting liquidation for position =>')} ${chalk.yellow(
+          'positionId:'
+        )} ${positionId} ${chalk.yellow('owner:')} ${owner} ${chalk.yellow(
+          'market:'
+        )} ${marketAddress}`
+      )
+
+      // attempt to liquidate
+      const liquidateTx = await marketContract.connect(wallet).liquidate(owner, positionId)
+      const receipt = await liquidateTx.wait()
+
+      log(
+        `${chalk.bgGreen('Position liquidated successfully! =>')} ${chalk.yellow(
+          'positionId:'
+        )} ${positionId} ${chalk.yellow('executed by:')} ${wallet.address} ${chalk.yellow(
+          'txHash:'
+        )} ${receipt.transactionHash}`
+      )
+
+      // remove position from Redis
+      await redis.hdel(`positions:${marketAddress}`, positionId)
+      await redis.zrem(`position_index:${marketAddress}`, positionId)
+
+      return
+    }
+
+    log(chalk.bold.red('No private keys available to liquidate position'))
+  } catch (error) {
+    log(chalk.bgRed('Error liquidating position:'), error)
+    log(chalk.bgBlue('Retrying in 10 seconds...'))
+    // simple retry mechanism
+    await new Promise((resolve) => setTimeout(resolve, 10000))
+    await liquidatePosition(position)
+  }
+}
+
+async function liquidablePositionsListener() {
+  if (!process.env.RPC_URLS) {
+    log(chalk.bold.red('At least one RPC_URLS must be provided'))
+    return
+  }
+
+  if (!process.env.PRIVATE_KEYS) {
+    log(chalk.bold.red('At least one private key is required'))
+    return
+  }
+
+  log(chalk.bold.blue('Executor started at:', new Date().toLocaleString()))
+  log(chalk.blue('Listening for liquidatable positions...'))
+
+  while (true) {
+    try {
+      const result = await redis.brpop('liquidatable_positions', 0)
+      if (result) {
+        const [key, positionData] = result
+        const position: Position = JSON.parse(positionData)
+
+        log(chalk.bold.blue('New liquidatable position found!'))
+        log(
+          chalk.blue(
+            `Processing position: marketAddress: ${position.marketAddress}, positionId: ${position.positionId}`
+          )
+        )
+
+        await liquidatePosition(position)
```

Comment:
Seems like this is waiting for one liquidation to finish before checking if there are more positions to be liquidated.
I think this is too slow, we should be able to liquidate positions simultaneously.

Reply:
Im agree. I modified the code to remove this re-try mechanism. Instead if the transaction fails, it will be re-added to the queue and it will have a maximum of 3 attempts to be processed.